src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs

```diff
@@ -157,6 +163,75 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
+        {
+            if (!CanApplyFirstRunDataMigration())
+            {
+                return;
+            }
+
+            _logger.LogWarning(LocalizableStrings.GlobalSettingsTemplatePackageProvider_DataMigrationWarning);
+
+            string sourceFeedKey = "NuGetSource";
+            string packageIdKey = "PackageId";
+            string versionKey = "Version";
+            string ownersKey = "Owners";
+            string trustedKey = "Trusted";
+
+            var updatedPackages = new List<TemplatePackageData>();
+            var metadataDownloader = new NuGetApiPackageManager(_environmentSettings);
+            var packages = await GetInstalledTemplatePackagesAsync(cancellationToken).ConfigureAwait(false);
+
+            try
+            {
+                foreach (var package in packages)
+                {
+                    string? packageSource = string.Empty;
+                    package.Details?.TryGetValue(sourceFeedKey, out packageSource);
+
+                    string? identifier = string.Empty;
+                    package.Details?.TryGetValue(packageIdKey, out identifier);
+
+                    string? version = string.Empty;
+                    package.Details?.TryGetValue(versionKey, out version);
+
+                    if (!string.IsNullOrWhiteSpace(packageSource))
+                    {
+                        var updatedPackageMetadata = await metadataDownloader.GetPackageMetadataAsync(
+                            identifier,
+                            version,
+                            packageSource,
+                            CancellationToken.None).ConfigureAwait(false);
+
+                        var extendedPackageDetails = package.Details.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+                        extendedPackageDetails[trustedKey] = updatedPackageMetadata.Trusted.ToString() ?? false.ToString();
+                        extendedPackageDetails[ownersKey] = updatedPackageMetadata.Owners;
+                        updatedPackages.Add(new TemplatePackageData(package.InstallerId, package.MountPointUri, package.LastChangeTime, extendedPackageDetails));
+                    }
+                    else
+                    {
+                        updatedPackages.Add(package);
+                    }
+                }
+
+                using var stream = _environmentSettings.Host.FileSystem.CreateFile(_paths.FirstRunCookie);
+
+                // cleanup existing cache file
+                await SetInstalledTemplatePackagesAsync(new List<TemplatePackageData>(), cancellationToken).ConfigureAwait(false);
+                await SetInstalledTemplatePackagesAsync(updatedPackages, cancellationToken).ConfigureAwait(false);
+            }
+            catch
+            {
+                if (_environmentSettings.Host.FileSystem.FileExists(_paths.FirstRunCookie))
+                {
+                    _environmentSettings.Host.FileSystem.FileDelete(_paths.FirstRunCookie);
+                }
```

Comment:
if migration critically failed, the file should be written not to re-attempt it.